kafka/kafka-common/src/test/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperatorTest.java

```diff
@@ -0,0 +1,399 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.malhar.kafka;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestWatcher;
+import org.junit.runners.Parameterized;
+import org.slf4j.LoggerFactory;
+
+import org.apache.apex.malhar.lib.wal.FSWindowDataManager;
+
+import com.google.common.collect.Lists;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.DAG;
+import com.datatorrent.api.DefaultInputPort;
+import com.datatorrent.api.LocalMode;
+import com.datatorrent.common.util.BaseOperator;
+import com.datatorrent.stram.StramLocalCluster;
+
+/**
+ * A bunch of test to verify the input operator will be automatically partitioned
+ * per kafka partition This test is launching its
+ * own Kafka cluster.
+ */
+public abstract class AbstractKafkaInputOperatorTest
+{
+
+  private int totalBrokers = 0;
+
+  private String partition = null;
+
+  public String testName = "";
+  protected boolean hasMultiPartition = false;
+  protected boolean hasMultiCluster = false;
+
+  public static String APPLICATION_PATH = AbstractEmbeddedKafka.baseDir + File.separator + StramLocalCluster.class.getName() + File.separator;
+
+  public class KafkaTestInfo extends TestWatcher
+  {
+    public org.junit.runner.Description desc;
+
+    public String getDir()
+    {
+      String methodName = desc.getMethodName();
+      String className = desc.getClassName();
+      return "target/" + className + "/" + methodName + "/" + testName;
+    }
+
+    @Override
+    protected void starting(org.junit.runner.Description description)
+    {
+      this.desc = description;
+    }
+  }
+
+  public static AbstractEmbeddedKafka kafkaServer;
```

Comment:
Can't this be 

 @Rule
 public KafkaUnitRule kafkaUnitRule = new KafkaUnitRule(zkPort, brokerPort);

and if multiple servers are needed create multiple rules?

Reply:
Same comment as above.